### ISSUE TITLE
Suggestion: 1:1 size by default unless larger than screen size

### ISSUE THREAD

**Issue by ccrowles** [riyasy/FlyPhotos#20]
Currently FlyPhotos has a fit-to-screen approach, which makes small images look extremely bad and grainy.

Example: this 350x350 pixels image
![image](https://github.com/user-attachments/assets/005f4a0f-f63a-45f2-be78-26e68e28275f)

It would be nice if it would instead behave by defaulting to having the images at 1:1 (actual size, not zoomed) and only the images that would go over the screen borders would be resized in a fit-to screen approach. Picasa had that exact approach by default and it was extremely nice.

**Comment by riyasy** [comment]
Valid suggestion. Shall take up for implementation in a future version.

**Comment by Orchestructive** [comment]
+1 to this feature request, I have a 48" 4k monitor and fullscreening a 240px image is quite a silly thing

**Comment by ittroll** [comment]
+1 from me too. 1:1 if it will fit, or ~90% of screen resolution if not. Just as Picasa did.

**Comment by Skyyblaze** [comment]
I would also like to see this implemented.

**Comment by clowngiver** [comment]
+1, came here to ask for this feature too, it is the only thing separating this program from basically perfection.

**Comment by riyasy** [comment]
The reason this feature hasn’t been added yet is because I want to do it in a way that doesn’t slow **Fly Photos** down.  Right now, Fly opens images as fast as possible by showing whatever is available first.  For example, if you open a huge **50-megapixel RAW file**, Fly will often start with the small embedded thumbnail (so you can see the photo instantly and may not be knowing the full image resolution), instead of waiting for the whole file to load.  Only if you have **“zoom on startup”** turned on will Fly wait for the high-quality image, because zooming into a blurry thumbnail wouldn’t look good.  

---

### Current scaling logic
The second challenge is making sure photo zooming and switching feels smooth and consistent.  
Fly uses one unified system for scaling so the experience isn’t jarring when you move from one image to another.  

**Example with numbers:**  

- Fly window = **1000px tall**  (Ignore width for simplicity)
  - Open a **2000px photo** → shown at **50% size** (fits height)  
  - Switch to a **3000px photo** → shown at **33% size**  

- Resize Fly window to **500px tall**  
  - 2000px photo → **25% size**  
  - Switch to a 3000px photo → **16% size**  

- Zoom out the first photo by **50% more**  
  - 2000px photo → **12.5% size**  
  - Switch to a 3000px photo → **8.3% size**  

This **unified scaling** is what makes switching between photos smooth and predictable in Fly. I have seen Microsoft Photos struggling with scaling logic when we have a folder with varied formats and varied sizes)

---

### Why “1:1 for small photo” rule is tricky
If I add a special rule like *always show very small photos at 1:1*, things get confusing:  

- Say you zoom out the first photo by **50%**, then switch to a **240px photo**  
  - Should that tiny photo now shrink to **120px**?  
  - Or should it stay at its original size?  

- What happens if you had switched to the 240 px photo after **resizing the window** to half of its original size?  Then if we maximize the window again, what to do?

---

### My goal
I want to make sure that:  
- Zooming and panning always feel natural  
- Your zoom/pan choices are preserved between photos  (unless the reset toggle is on in Settings window)
- The experience stays fast and consistent

**Comment by ittroll** [comment]
Thanks for taking the time to explain the logic and challenges you are facing.

Is it possible to quickly read the image size (Dimensions) from the files properties (metadata) before loading in the whole file? You would then know in advance how to scale for display.

I actually find the switch from the low resolution image to the high on startup quite jarring as the full image pops in. So I do use the the "zoom on startup" option to prevent this, which mimics the Picasa behaviour. Thanks for including this! An option to wait for the full image on startup would work equally well. The low resolution thumbnails would only show when speed scrolling or skipping ahead.

I'm not sure how useful it is to maintain the zoom/pan choices when switching between images of different resolutions/ratios. It is unlikely to display in a desirable way. A prime example is  a folder with a mix of portrait and landscape orientation photos. Picasa initially displays all images at the default zoom level (90% or 1:1), but then remembers any subsequent custom zoom/pan for each individual image until the app is closed.

I can appreciate that maintaining a consistent zoom/pan may be useful if the images are all the same resolution. So, perhaps the logic or option could be "Reset pan and zoom when switching images with a different resolution"? I guess this would depend on reading the dimensions property in advance.

**Comment by ccrowles** [comment]
As ittroll said before, thanks for explaining in detail the challenges of this feature implementation. While I do not have technical suggestions, I do agree with ittroll that for my personal use I don't consider useful to mantain the zoom/pan choices between images. Picasa never did, and I was absolutely alright with it since usually in a folder I tend to have images with vastly different resolutions and ratios. Obviously I'm not everyone, so people's mileage may vary.

(And I also, just like ittroll, have the "zoom on startup" enabled because I find visually annoying the switch from low res to high res.)

**Comment by riyasy** [comment]
> Is it possible to quickly read the image size (Dimensions) from the files properties (metadata) before loading in the whole file? You would then know in advance how to scale for display.

This is the only way.. The way to read this may be different for each file type. (Especially for file types for which Fly uses its own preview maker for speed benefits instead of Windows imaging component). Will need exhaustive testing with different image formats.. each with and without embedded thumbnails etc etc.. Also will need to provision the width and height in disk cache for the recently opened ones. 

> I'm not sure how useful it is to maintain the zoom/pan choices when switching between images of different resolutions/ratios. It is unlikely to display in a desirable way. 

Actually this was created for myself at start before open sourcing and I liked it that way since I mostly use this to view my Camera photos. Then later added the reset button in Settings.


As a start I am trying to get the width, height of the image in the fastest way possible and use it to scale the preview. Otherwise the lowres image will be shown as small and in background when I load the full res image, it will fill the screen. You can occasionally see these types of behaviours in Microsoft Photos when we open a new folder which had not been cached.

I need to scale the low-res and show it in the same exact box of what the high res image is going to be loaded and shown in the future so that the transition seems as smooth as possible. 
Still jarring can happen because for many files, the embedded lowres image can be slightly different than the actual image even in aspect ratio. For e.g in my Sony Camera RAW, there are black bars on top and bottom in the embedded thumbnail which causes a jarring during switch to highres which I won't be able so solve.

**Comment by riyasy** [comment]
> As ittroll said before, thanks for explaining in detail the challenges of this feature implementation. While I do not have technical suggestions, I do agree with ittroll that for my personal use I don't consider useful to mantain the zoom/pan choices between images. Picasa never did, and I was absolutely alright with it since usually in a folder I tend to have images with vastly different resolutions and ratios. Obviously I'm not everyone, so people's mileage may vary.
> 
> (And I also, just like ittroll, have the "zoom on startup" enabled because I find visually annoying the switch from low res to high res.)

Yes.. I understand that many don't like preserving the pan,zoom across images. Actually this was created for myself at start before open sourcing and I liked it that way since I mostly use this to view my Camera photos. Then later added the reset button in Settings. Currently the default behaviour is to preserve, may be we can change the default behaviour in a future release.

**Comment by ittroll** [comment]
> Still jarring can happen because for many files, the embedded lowres image can be slightly different than the actual image even in aspect ratio. For e.g in my Sony Camera RAW, there are black bars on top and bottom in the embedded thumbnail which causes a jarring during switch to highres which I won't be able so solve.

I also have a Sony camera. In the case of RAW files, I think the low-res thumbnail is based on camera jpg. So when the actual full-res image loads it has uncorrected brightness and lens distortion. This exaggerates the popping effect as the image changes significantly. But this is avoided with the zoom on startup option.

Anyway keep up the good the work!

**Comment by clowngiver** [comment]
I don't have anything to add on the technical aspect of the conversation, but wanted thank you for being open to suggestions, even with the caveat of the perma-fit this is the first viable replacement to Picasa I've found in like 10 years, surpasses it in some aspects (like actually displaying very large images), so yeah thanks a bunch

**Comment by riyasy** [comment]
Please find the new installer for 2.5.4 Beta which I am sharing here. This has the proposed 1:1 image fix.
Please delete %LOCALAPPDATA%\FlyPhotos before installing.
Also uninstall previous versions if any
https://drive.google.com/file/d/1D9NNrpYpNXguJ-NcfkKXQoJbfYOYtoOM/view?usp=sharing

Images if greater than the application screen area (can be lesser depending on usable % set in Settings window) is zoomed out to fit. If images are smaller than this area, they are shown as 1:1. Check window maximize and resize behaviour also and check inconsistencies if found.

@ittroll has reported he has issues with 2.5.3 crashing when we take settings page. I haven't done anything particular to fix that. May be the issue carries forward to 2.5.4 on ittroll's machine.

**Comment by ittroll** [comment]
This is working beautifully for me. I've tested with JPG, GIF, PNG, HEIC, and ARW. Also loving the % screen usage option.

~In 2.5.2 I think it was possible to set the UI fade to 0, so that the UI completely disappears (like Picasa). However, in this release the lowest is 30. Is there a reason for the change or am I misremembering?~
Edit: Ignore the above. I of course need to set fade to 100 to make it disappear. The setting is for fade intensity not visibility. There is a typo in the UI as it is labelled intesity rather than intensity.

I've updated the crashing issue - seems this was due to me attempting an in-place upgrade.

**Comment by clowngiver** [comment]
Amazing, just what I hoped for, tried it for a couple hours in a folder with mixed file types and sizes, and I noticed no issues with the resizing or slow-downs of any type. 

When zooming onto smaller objects, if the mouse isn't on the picture it'll just zoom in into empty space which confused me at first, but on second glance that's a side effect of how zooming works in the application rather than something to do with this update, it'll just zoom where your mouse is, which is in line with the original use of the program (photos that cover the entire screen)

Very happy honestly, many thanks! Ah one last thing, while testing it, I noticed that in the context menu cut/copy don't work, I thought maybe it was menu wide, but everything else worked fine, just those two, no idea if it was this update that broke it or if it is an ongoing issue since I never tried it before.

**Comment by riyasy** [comment]
> Very happy honestly, many thanks! Ah one last thing, while testing it, I noticed that in the context menu cut/copy don't work, I thought maybe it was menu wide, but everything else worked fine, just those two, no idea if it was this update that broke it or if it is an ongoing issue since I never tried it before.

Yes.. the context menu integration is minimal. I have somehow pulled up and shown the windows explorer context menu, but many menu items may not work properly. Please raise a new ticket. We can track it there.

**Comment by ittroll** [comment]
I have noticed that PNG screenshots don't display clearly when viewed 1:1. If interpolation is on, text has a slight blurriness. If interpolation is off, text breaks up quite badly. The effect is most noticeable when the background in white.

I guess this is down to the image renderer. Picasa must use a different method as the text appears crisp.

<img width="650" height="470" alt="Image" src="https://github.com/user-attachments/assets/395f5f78-7056-4c26-94cb-6a0f40e38654" />

**Comment by riyasy** [comment]
> I have noticed that PNG screenshots don't display clearly when viewed 1:1. If interpolation is on, text has a slight blurriness. If interpolation is off, text breaks up quite badly. The effect is most noticeable when the background in white.
> 
> I guess this is down to the image renderer. Picasa must use a different method as the text appears crisp.
> 

Can you try with the Checkered background off..  ?

#42 , To fix this bug, I had turned off antialiasing in PNG and other formats which support transparency.

**Comment by ittroll** [comment]
> Can you try with the Checkered background off.. ?

That option was already off. But I tried with it both on and off and it makes no difference.
The same issue affects the image even when converted to other formats; GIF, BMP, JPG.

Here is the original screenshot image.
<img width="449" height="287" alt="Image" src="https://github.com/user-attachments/assets/9b533029-fef2-4f98-8160-6f188b970b58" />

Here is a screenshot of that image when viewed with FlyPhotos (with interpolation off) on my machine.
<img width="449" height="287" alt="Image" src="https://github.com/user-attachments/assets/6d8f64f3-3eae-4080-8687-36edf746a73a" />

The dimensions are correct for a 1:1 display, but it looks like it has been somehow resized, or some extra anti-aliasing is being applied. What is really strange is the level of distortion varies as I move the image around the screen. Like it is being re-rendered differently for the new screen position.

**Comment by riyasy** [comment]
Ok.. I shall check..

**Comment by riyasy** [comment]
On my windows 11 pc, I see Fly rendering without distortion. (Edit. checked now in Windows 10 PC also)

<img width="579" height="844" alt="Image" src="https://github.com/user-attachments/assets/6fa2a19e-9fe9-4d7c-824f-974d388375bf" />

**Comment by riyasy** [comment]
@ittroll, Is your display > scaling settings set to 100 %?  If you don't have the issue at 100% and has the issue only at non 100%, please raise as a new issue.

**Comment by ittroll** [comment]
Thanks for testing. Display is set to 100% scaling, resolution 3840 x 2160 (4K), Nvidia GPU, Windows 11 24H2.
If no one else reports the issue, then it must be something specific to my system. Odd that only FlyPhotos is affected though.

Are you using a particular library or method the display the image? Does it have any dependencies (I have uninstalled Microsoft Photos)? Knowing this will help me to research the problem.

**Comment by riyasy** [comment]
@ittroll , please continue the discussion in https://github.com/riyasy/FlyPhotos/issues/55

**Comment by ittroll** [comment]
Will do. Feel free to delete the last few posts in this thread if you want to clean it up.

**Comment by riyasy** [comment]
Closed with v2.5.4 release.
